pulsar-client/src/main/java/org/apache/pulsar/client/impl/TransactionMetaStoreHandler.java

```diff
@@ -170,17 +162,26 @@ public void connectionOpened(ClientCnx cnx) {
                     return null;
                 });
             } else {
-                if (!changeToReadyState()) {
-                    cnx.channel().close();
-                } else {
-                    connectionHandler.setClientCnx(cnx);
-                    cnx.registerTransactionMetaStoreHandler(transactionCoordinatorId, this);
-                }
-                this.connectFuture.complete(null);
+                registerToConnection(cnx);
             }
         });
     }
 
+    private boolean registerToConnection(ClientCnx cnx) {
+        if (changeToReadyState()) {
+            connectionHandler.setClientCnx(cnx);
+            cnx.registerTransactionMetaStoreHandler(transactionCoordinatorId, this);
+            connectFuture.complete(null);
+            return true;
+        } else {
+            State state = getState();
+            cnx.channel().close();
+            connectFuture.completeExceptionally(
+                    new IllegalStateException("Failed to change the state from " + state + " to Ready"));
+            return false;
```

Comment:
Did you want to explicitly set the handler state to closed? I see there was a call to `setState(State.Closed)` - I'm not sure if it is called implicitly or there is a missing unit test that should've failed.

Reply:
The failure of  `changeToReadyState` is an unexpected behavior, which means the state is not `Uninitialized`, `Connecting`, `Ready` or `RegisteringSchema`, i.e. the state could be one of the following possible values:
- `Closing`
- `Closed`
- `Terminated`
- `Failed`
- `ProducerFenced`

The failure might be caused by the `TransactionMetaStoreHandler` was closed or encountered some errors somewhere before the connection has not completed.

I think the previous call to `setState(State.Closed)` does not make sense. It could hide the actual error that makes `changeToReadyState` fail.

Reply:
Thanks for explaining! I assumed that the Handler state was meant to mirror the underling channel state - I'm new to this code but with the lack of a unit test it is hard to tell what is the intended behavior.

Would it be possible to add a test that verifies there are no side effects with regards to state when `changeToReadyState` fails?

Reply:
> Handler state was meant to mirror the underling channel state

There is no state (other than closed or not closed) of a Netty channel. We should not map states like Terminated to "closed".

BTW, I've added the unit test, PTAL.